### PR TITLE
DISTINCT in order optimization

### PR DESCRIPTION
**Pull request by devcrafter**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Distinct optimization for sorted columns. Use specialized distinct transformation in case input stream is sorted by column(s) in distinct. Optimization can be applied to pre-distinct, final distinct, or both. Initial implementation by @dimarub2000 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Let's also reference the original author and the original pull request in the changelog (we have this practice if the original author has done more than 0% of the work).

**Comment by devcrafter** [comment]
@nickitat please could you review it?

**Comment by nickitat** [comment]
notes regarding the approach:
* distinct usually happens in two stages: multi-threaded pre-distinct and single-threaded final distinct. and it doesn't make much practical sense to optimize only pre-distinct, because, in case of high-cardinality data, execution time will be dominated by final distinct. so, IMO, if optimization is applicable on the final stage - we should apply it.
* on the road to DISTINCT transform, data may be already reordered. in case of pre-distinct it may happen in query like `select distinct n from t group by n`, in case of final distinct - in query like `select distinct n from t order by sipHash64(n)`. probably we should look into `input_streams.front().sort_description` in DistinctStep when adding transforms into pipeline. 
* to read data initially in-order we need to choose the special type of reading for that as we do [for aggregation](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L978-L980). and propagate info about sorting order through `sort_description` [here](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L82-L86).

**Comment by devcrafter** [comment]
@Mergifyio update

**Comment by devcrafter** [comment]
@Mergifyio update

**Comment by devcrafter** [comment]
@Mergifyio update

**Comment by devcrafter** [comment]
@nickitat Please take a look. Thanks

File: tests/queries/0_stateless/02317_distinct_in_order_optimization_explain.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+[ ! -z "$CLICKHOUSE_CLIENT_REDEFINED" ] && CLICKHOUSE_CLIENT=$CLICKHOUSE_CLIENT_REDEFINED
+
+DISABLE_OPTIMIZATION="set optimize_distinct_in_order=0"
+ENABLE_OPTIMIZATION="set optimize_distinct_in_order=1"
+GREP_OPTIMIZATIONS="grep 'DistinctSortedChunkTransform\|DistinctSortedTransform'"
```
Comment (devcrafter) [inline tests/queries/0_stateless/02317_distinct_in_order_optimization_explain.sh:11]: `DistinctSortedTransform` need to be deleted here.

File: tests/queries/0_stateless/02317_distinct_in_order_optimization_explain.reference [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+DistinctSortedChunkTransform
+-- distinct with primary key prefix and order by on column in distinct -> pre-distinct and final distinct optimization
+DistinctSortedChunkTransform
+DistinctSortedChunkTransform
+-- distinct with primary key prefix and order by on column _not_ in distinct -> pre-distinct optimization only
+DistinctSortedChunkTransform
+-- distinct with non-primary key prefix -> no optimizations
+No optimizations
+-- distinct with non-primary key prefix and order by on column in distinct -> final distinct optimization only
+DistinctSortedChunkTransform
```
Comment (devcrafter) [inline tests/queries/0_stateless/02317_distinct_in_order_optimization_explain.reference:17]: I'd prefer to have a different name for final distinct, i.e. `DistinctSortedStreamTransform`. But the implementation is the same, so I'm not sure if it is worth the hassle to have a flag (i.e.`final_distinct`), and provide the different name if it's true. @KochetovNicolai any thoughts?

**Comment by devcrafter** [comment]
@mergifyio update

File: src/Processors/Transforms/DistinctSortedChunkTransform.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+#include <Processors/ISimpleTransform.h>
+
+namespace DB
+{
+
+///
+/// DISTINCT optimization for sorted chunks
+/// It also checks if current chunks is continuation of previous one,
+/// i.e. sorting prefix value of last row in previous chunk is the same as of first row in current one,
+/// so it can correctly process sorted stream as well
```
Comment (nickitat) [inline src/Processors/Transforms/DistinctSortedChunkTransform.h:16]: it worth to mention here the logic depending on `other_columns.empty()`. if true, we will go without hash tables at all. if false - HT will be created for each value of sorting key prefix which means lower memory consumption.

**Review by nickitat — COMMENTED** [review]
we also need perf tests. with low or high cardinality data x only pre-distinct or both x with or without ordinary distinct under the hood

**Comment by devcrafter** [comment]
@Mergifyio update

**Comment by devcrafter** [comment]
@Mergifyio update

File: src/Processors/Transforms/DistinctSortedChunkTransform.cpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    initChunkProcessing(input_columns);
+
+    /// build filter:
+    /// (1) find range with the same values in sorted columns -> [range_begin, range_end)
+    /// (2) for found range
+    ///     if there is no "non-sorted" columns: filter out all rows in range except first one
+    ///     otherwise: apply ordinary distinct
+    /// (3) repeat until chunk is processed
+    IColumn::Filter filter(chunk_rows);
+    auto [range_begin, output_rows] = continueWithPrevRange(chunk_rows, filter); /// try to process chuck as continuation of previous one
```
Comment (nickitat) [inline src/Processors/Transforms/DistinctSortedChunkTransform.cpp:190]: I suggest to rename `range_begin` to `next_range_begin`

**Comment by nickitat** [comment]
perf tests results look really good.
I think speed up might be even bigger if we apply read in order. but it makes sense to do it out of scope of this pr.

**Comment by devcrafter** [comment]
@Mergifyio update

File: src/Processors/Transforms/DistinctSortedChunkTransform.cpp [hunk 5]
```diff
@@ -123,10 +122,23 @@ bool DistinctSortedChunkTransform::isCurrentKey(const size_t row_pos) const
...
+    if (linear_probe_end > end)
+        linear_probe_end = end;
+
+    for (size_t pos = begin; pos < linear_probe_end; ++pos)
+    {
+        if (!isCurrentKey(pos))
+            return pos;
+    }
+
+    size_t low = linear_probe_end;
```
[the remark this answers is not in the document]
Comment (devcrafter) [reply src/Processors/Transforms/DistinctSortedChunkTransform.cpp:140]: I'd check it in a follow-up PR

**Pull request merged**